fabric-tag-api-v1/src/main/java/net/fabricmc/fabric/impl/tag/TagAliasLoader.java

```diff
@@ -0,0 +1,176 @@
+
+			for (Map.Entry<Identifier, Resource> entry : resourceFinder.findResources(manager).entrySet()) {
+				Identifier resourcePath = entry.getKey();
+				Identifier groupId = resourceFinder.toResourceId(resourcePath);
+
+				try (Reader reader = entry.getValue().getReader()) {
+					JsonElement json = JsonParser.parseReader(reader);
+					Codec<TagAliasGroup<Object>> codec = TagAliasGroup.codec((RegistryKey<? extends Registry<Object>>) registryKey);
+
+					switch (codec.parse(JsonOps.INSTANCE, json)) {
+					case DataResult.Success(TagAliasGroup<Object> group, Lifecycle unused) -> {
+						var data = new Data(groupId, group);
+						dataByRegistry.computeIfAbsent(registryKey, key -> new ArrayList<>()).add(data);
+					}
+					case DataResult.Error<TagAliasGroup<Object>> error -> {
+						LOGGER.error("[Fabric] Couldn't parse tag alias group file '{}' from '{}': {}", groupId, resourcePath, error.message());
+					}
+					}
+				} catch (IOException | JsonParseException e) {
+					LOGGER.error("[Fabric] Couldn't parse tag alias group file '{}' from '{}'", groupId, resourcePath, e);
+				}
+			}
+		}
+
+		return dataByRegistry;
+	}
+
+	private static String getDirectory(RegistryKey<? extends Registry<?>> registryKey) {
+		String directory = "fabric/tag_alias/";
+		Identifier registryId = registryKey.getValue();
+
+		if (!Identifier.DEFAULT_NAMESPACE.equals(registryId.getNamespace())) {
+			directory += registryId.getNamespace() + '/';
+		}
+
+		return directory + registryId.getPath();
+	}
+
+	@Override
+	protected void apply(Map<RegistryKey<? extends Registry<?>>, List<TagAliasLoader.Data>> prepared, ResourceManager manager, Profiler profiler) {
+		for (Map.Entry<RegistryKey<? extends Registry<?>>, List<Data>> entry : prepared.entrySet()) {
+			Map<TagKey<?>, Set<TagKey<?>>> groupsByTag = new HashMap<>();
+
+			for (Data data : entry.getValue()) {
+				Set<TagKey<?>> group = new HashSet<>(data.group.tags());
+
+				for (TagKey<?> tag : data.group.tags()) {
+					Set<TagKey<?>> oldGroup = groupsByTag.get(tag);
+
+					// If there's an old group...
+					if (oldGroup != null) {
+						// ...merge all of its tags into the current group...
+						group.addAll(oldGroup);
+
+						// ...and replace the recorded group of each tag in the old group with the new group.
+						for (TagKey<?> other : oldGroup) {
+							groupsByTag.put(other, group);
+						}
+					}
+
+					groupsByTag.put(tag, group);
+				}
+			}
+
+			// Remove any groups of one tag, we don't need to apply them.
+			groupsByTag.values().removeIf(tags -> tags.size() == 1);
+
+			RegistryWrapper.Impl<?> wrapper = registries.getOrThrow(entry.getKey());
+
+			if (wrapper instanceof TagAliasEnabledRegistryWrapper aliasWrapper) {
+				aliasWrapper.fabric_loadTagAliases(groupsByTag);
+			} else {
+				LOGGER.error("[Fabric] Couldn't apply tag aliases to registry wrapper {} ({}) since it had an unknown type", wrapper, entry.getKey().getValue());
```

Comment:
Nit: This sounds like a fatal error to me. If it happens its likely an issue with fabric or another mod and not the datapack.

Reply:
Agreed, but I don't think throwing makes much sense here since it can put resource reloading in a loop.

Reply:
(The infinite loop is for resource packs only, and I am not sure if that happens in 1.21.4. There's some refactor around this code every other version or so...)